File: dotnet/GameStateTest/ObserverTest.cs
using GameLib.Net.Game.State;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameStateTest
{
    [TestClass]
    public class ObserverTest
    {
        private const string STATE_ID = "State";

        private Root root;
        private BooleanState state;
        private IObserver observer;

        [TestInitialize]
        public void SetUp()
        {
            observer = Mock.Of<IObserver>();
            root = StateTestUtils.SetUpRoot();
            state = BooleanState.Create(root, STATE_ID, false);
            state.AddObserver(observer);
        }

        [TestMethod]
        public void TestUpdate()
        {
            Assert.IsTrue(state.GetObservers().Contains(observer));
            state.Set(true);
            Mock.Get(observer).Verify(observer => observer.Update(state.ToText()), Times.Never());
            StateTestUtils.Close(root);
            Mock.Get(observer).Verify(observer => observer.Update(state.ToText()), Times.Once());
        }
    }
}

File: src/test/java/net/sf/rails/game/state/ObserverTest.java
package net.sf.rails.game.state;

import static org.mockito.Mockito.*;
import static org.fest.assertions.api.Assertions.assertThat;

import net.sf.rails.game.state.BooleanState;
import net.sf.rails.game.state.Observer;
import net.sf.rails.game.state.Root;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

@RunWith(MockitoJUnitRunner.class)
public class ObserverTest {
    
    private final static String STATE_ID = "State";
    
    private Root root;
    private BooleanState state;
    @Mock private Observer observer;

    @Before
    public void setUp() {
        root = StateTestUtils.setUpRoot();
        state = new BooleanState(root, STATE_ID, false);
        state.addObserver(observer);
    }
    
    @Test
    public void testUpdate() {
        assertThat(state.getObservers()).contains(observer);
        state.set(true);
        verify(observer, never()).update(state.toText());
        StateTestUtils.close(root);
        verify(observer).update(state.toText());
    }

}
